packages/core/custom-branding/core-custom-branding-server-internal/custom_branding_service.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import type {
+  CustomBrandingStart,
+  CustomBrandingStartDeps,
+} from '@kbn/core-custom-branding-server';
+import type { CustomBranding } from '@kbn/core-custom-branding-common';
+import { KibanaRequest } from '@kbn/core-http-server';
+import { InternalSavedObjectsServiceStart } from '@kbn/core-saved-objects-server-internal';
+import { InternalUiSettingsServiceStart } from '@kbn/core-ui-settings-server-internal';
+import type { CoreContext } from '@kbn/core-base-server-internal';
+import type { Logger } from '@kbn/logging';
+import { IUiSettingsClient } from '@kbn/core-ui-settings-server';
+
+/**
+ * @internal
+ */
+export interface InternalCustomBrandingSetup {
+  register: (pluginName: string) => void;
+}
+
+export class CustomBrandingService {
+  private pluginName: string | undefined;
+  private savedObjects?: InternalSavedObjectsServiceStart;
+  private uiSettings?: InternalUiSettingsServiceStart;
+  private readonly settingsKeys: Array<keyof CustomBranding> = [
+    'logo',
+    'customizedLogo',
+    'faviconPNG',
+    'faviconSVG',
+    'pageTitle',
+  ];
+  private logger: Logger;
+
+  constructor(coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('custom-branding-service');
+  }
+
+  private getBrandingFrom = async (uiSettingsClient: IUiSettingsClient) => {
+    const branding: CustomBranding = {};
+    for (let i = 0; i < this.settingsKeys!.length; i++) {
+      const key = this.settingsKeys[i];
+      const fullKey = `customBranding:${key}`;
+      const value = await uiSettingsClient.get(fullKey);
+      if (value) {
+        branding[key] = value;
+      }
+    }
+    return branding;
+  };
+
+  public getBrandingFor = async (request: KibanaRequest): Promise<CustomBranding> => {
+    if (!this.pluginName || this.pluginName !== 'customBranding' || !this.settingsKeys) {
+      return {};
+    }
+    const soClient = this.savedObjects!.getScopedClient(request);
+    const uiSettings = this.uiSettings!.globalAsScopedToClient(soClient);
+    return await this.getBrandingFrom(uiSettings);
+  };
```

Comment:
This shouldn't be exposed directly from the service, and the service shouldn't be propagated to the rendering service by itself.

Instead, this API should be exposed via the `InternalCustomBrandingSetup` contract, and returned by the service's setup method. 

We should also add a sanity check in the method, just to make sure it isn't called before #start was called.

Something as naive as

```ts
if(!this.savedObjects || !this.uiSettings) {
    throw new Error('getBrandingFor cannot be called before #start')
}
```

would do the trick

Reply:
I am confused by this flow; that's why I added it as a standalone-method in the first place. We're exposing the method from the *setup*, but it cannot be called before *start* was called. Is there anywhere else where we're violating the lifecycle in this way?